### PR TITLE
benchmark: add initial documentation

### PR DESCRIPTION
**Pull request by igoropaniuk**
Add initial benchmark framework documentation. Describe implementation
details, usage, current limitations and further steps (what should be added in the nearest future)

`Signed-off-by: Igor Opaniuk <igor.opaniuk@linaro.org>`

### CODE REVIEW

**Comment by jforissier** [comment]
We need the source files for the `.png` diagrams.

**Comment by igoropaniuk** [comment]
@jforissier added source files and a comment in the `benchmark.md` which describes how to edit them

File: documentation/benchmark.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+4. [Limitations and further steps](#4-limitations-and-further-steps)
+
+
+## 1. Introduction
+Due to its nature, OP-TEE is being a solution spanning over several
+architectural layers, where each layer includes its own complex parts.
+For further optimizations of performance, there is a need of tool which will
+provide detailed and precise profiling information for each layer.
+
+It's necessary to receive measurements of latency for:
```
Comment (etienne-lms) [inline documentation/benchmark.md]: minor: prefer `It is` than `It's`. Avoid contractions when not really needed.

File: documentation/benchmark.md [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+* The roundtrip time for going from a client application in normal world,
+down to a Trusted Application and back again.
+* Detailed information for amount of time taken to go through each layer:
+	* libTEEC -> OP-TEE KMOD
+	* OP-TEE KMOD -> OP-TEE OS Core
+	* OP-TEE OS Core -> LibUTEE
```
Comment (etienne-lms) [inline documentation/benchmark.md]: I would replace `LibUTEE` here by `TA entry point` (without specifying wherever it is).

File: documentation/benchmark.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+### 2.1 Design overview
+
+Benchmark framework consists of such components:
+1. **Benchmark CA**: a dedicated client application, which is responsible
+for allocating timestamp circular buffers, registering these buffers in
+the **Benchmark PTA**, and consuming all timestamp data, generated
+by all OP-TEE layers. Finally, it puts timestamp data into appropriate
+file with `.ts` extension. Additional details can be found here
+here [optee_benchmark]
+2. **Benchmark TA**: pseudo TA, which owns all per-cpu circular buffers.
```
Comment (etienne-lms) [inline documentation/benchmark.md]: "Benchmark **P**TA"

File: documentation/benchmark.md [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+**Benchmark CA** performs registration of all buffers and stores memrefs in
+this PTA. Also, **Benchmarks PTA** is reponsible for registering timestamp
+buffers in other OP-TEE layers (**libTEEC** and
+**Linux OP-TEE kernel driver**).
```
Comment (etienne-lms) [inline documentation/benchmark.md]: Typo: replace "libTEEC" and "Linux OP-TEE kernel driver" with OPTEE "core" and OPTEE "libutee".

File: documentation/benchmark.md [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+3. There are also benchmark components in the **libTEEC** and **Linux OP-TEE
+kernel driver**, that mainly handles timestamp buffer registration requests.
+
+Also, each layer has ability to store timestamps to the timestamp
+circular buffers.
```
Comment (etienne-lms) [inline documentation/benchmark.md]: I would replace with: `When "Benchmark is enabled, all OP-TEE layers (libteec, kernel driver, core, libutee) do fill the registered timestamp circular buffer with timestamp data for all invocation requests.`

File: documentation/benchmark.md [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+![design_overview][design_overview]
+
+### 2.2 Timestamp source
+
+ARM Performance Monitor Units are used as the main source of timestamp values.
+The reason why this technology was chosen is that it's supported on all
+ARMv7/ARMv8 cores. Besides it can provide precise pre-cpu cycle counter values,
+it's possible to enable EL0 access to all events, so usermode applications can
+directly read cpu counter values from coprocessor registers,
+achieving miniman latency by avoiding additional syscalls to EL1 core.
```
Comment (etienne-lms) [inline documentation/benchmark.md]: s/miniman/minima/

File: documentation/benchmark.md [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+Besides CPU cycle counter values, timestamp by itself contains also
+information about:
+* CPU core id
```
Comment (etienne-lms) [inline documentation/benchmark.md]: `executing CPU core index` ?

File: documentation/benchmark.md [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+* OP-TEE layer id, where this timestamp was
+obtained from
+* Memory address, which can be used for getting a symbol name
```
Comment (etienne-lms) [inline documentation/benchmark.md]: `Program counter content when timestamp was logged, which can be used ...`

File: documentation/benchmark.md [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+stored.
+
+### 3.2 Adding custom timestamps
+Currently, timestamping is done only for `InvokeCommand` calls, but it's also
+possible to choose custom places in the supported OP-TEE layers.
+
+To add timestamp storing command to custom c source file:
+1. Include appropriate header:
+	* OP-TEE OS Core: `bench.h`
+    * Linux OP-TEE kmod: `optee_bench.h`
```
Comment (etienne-lms) [inline documentation/benchmark.md]: indentation

File: documentation/benchmark.md (hunk 4 above)
Comment (igoropaniuk) [inline documentation/benchmark.md]: Didn't get this one
maybe you meant:
libTEEC -> OPTEE "libteec"
Linux OP-TEE kernel driver -> OP-TEE kernel driver
?

Comment (etienne-lms) [inline documentation/benchmark.md]: Maybe the paragraph should be rephrased:
```
2. **Benchmark PTA**: pseudo TA, which owns all per-cpu circular buffers.
   **Benchmark PTA** must be invoked (by a CA) to register the time stamp circular buffers. In turn,
   the pTA invokes the optee linux driver (through some RPC mean) to register this circular buffers
   in the linux kernel layer.

3. **libTEEC** and **Linux OP-TEE kernel driver** include means to store timestamps in the 
   circular buffers once  these are registered by the **Banchmark pTA** into the "Linux OP-TEE
   kernel driver".
```

Comment (igoropaniuk) [inline documentation/benchmark.md]: @etienne-lms 
rephrased, thanks for your suggestion

**Comment by igoropaniuk** [comment]
@etienne-lms all comments addressed, thanks!

File: documentation/benchmark.md [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+4. [Limitations and further steps](#4-limitations-and-further-steps)
+
+
+## 1. Introduction
+Due to its nature, OP-TEE is being a solution spanning over several
+architectural layers, where each layer includes its own complex parts.
+For further optimizations of performance, there is a need of tool which will
+provide detailed and precise profiling information for each layer.
+
+It is necessary to receive measurements of latency for:
```
Comment (etienne-lms) [inline documentation/benchmark.md]: This sentence ending with each of the 2 bullets song strange not my (french) ears. Can one cross check?

File: documentation/benchmark.md [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+	* OP-TEE OS Core -> TA entry point (**not supported yet**)
+	* The same way back
+---
+## 2. Implementation details
+### 2.1 Design overview
+
+Benchmark framework consists of such components:
+1. **Benchmark CA**: a dedicated client application, which is responsible
+for allocating timestamp circular buffers, registering these buffers in
+the **Benchmark PTA**, and consuming all timestamp data, generated
```
Comment (etienne-lms) [inline documentation/benchmark.md]: minor: remove the coma: `.. all timestamp data generated by ...`
`

File: documentation/benchmark.md [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+by all OP-TEE layers. Finally, it puts timestamp data into appropriate
+file with `.ts` extension. Additional details can be found here
+here [optee_benchmark]
+2. **Benchmark PTA**: pseudo TA, which owns all per-cpu circular buffers.
```
Comment (etienne-lms) [inline documentation/benchmark.md]: i would add "non secure", i.e "... which owns all per-cpu circular **non secure** buffers **from a shared memory**.".

File: documentation/benchmark.md [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+**Benchmark PTA** must be invoked (by a CA) to register the time stamp
```
Comment (etienne-lms) [inline documentation/benchmark.md]: s/time stamp/timestamp/ for consistency.

File: documentation/benchmark.md [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+circular buffers. In turn, the pTA invokes the optee linux driver
```
Comment (etienne-lms) [inline documentation/benchmark.md]: s/pTA/**Benchmark PTA**/ for consistency.

File: documentation/benchmark.md [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+(through some RPC mean) to register this circular buffers in the linux
+kernel layer.
+3. **libTEEC** and **Linux OP-TEE kernel driver** include means to store
+timestamps in the circular buffers once these are registered by the
+**Banchmark pTA** into the "Linux OP-TEE kernel driver".
```
Comment (etienne-lms) [inline documentation/benchmark.md]: Bold "Linux OP-TEE kernel driver" for consistency.

File: documentation/benchmark.md [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+
+Benchmark is enabled, all OP-TEE layers (libteec, kernel driver, core)
```
Comment (etienne-lms) [inline documentation/benchmark.md]: "**When b** ~~B~~enchmark is enabled `(CFG_TEE_BENCHMARK=y)`, all OP-TEE layers ..."

File: documentation/benchmark.md [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+do fill the registered timestamp circular buffer with timestamp data for
+all invocation requests
```
Comment (etienne-lms) [inline documentation/benchmark.md]: Add terminal dot.

Suggest adding this: `... for all invocation requests on condition that the circular buffer is allocated/registered.`

File: documentation/benchmark.md [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+
+![design_overview][design_overview]
+
+### 2.2 Timestamp source
+
+ARM Performance Monitor Units are used as the main source of timestamp values.
+The reason why this technology was chosen is that it's supported on all
```
Comment (etienne-lms) [inline documentation/benchmark.md]: s/it's/it is/

File: documentation/benchmark.md [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+ARMv7/ARMv8 cores. Besides it can provide precise pre-cpu cycle counter values,
+it's possible to enable EL0 access to all events, so usermode applications can
```
Comment (etienne-lms) [inline documentation/benchmark.md]: s/it's/it is/

File: documentation/benchmark.md [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+information about:
+* Executing CPU core index
+* OP-TEE layer id, where this timestamp was
+obtained from
+* Program counter value when timestamp was logged, which can be used for
+getting a symbol name (a filename and line number)
+
+### 2.3 Call sequence diagram
+
+![call_sequence][call_sequence]
```
Comment (etienne-lms) [inline documentation/benchmark.md]: "1. Allocate per-cpu timestamp ringbuffer":
The allocation request goes to the linux kernel optee driver where allocation is done.
Am I right? The arrow goes up to TEE.

File: documentation/benchmark.md [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+Besides CPU cycle counter values, timestamp by itself contains also
+information about:
+* Executing CPU core index
+* OP-TEE layer id, where this timestamp was
+obtained from
+* Program counter value when timestamp was logged, which can be used for
+getting a symbol name (a filename and line number)
+
+### 2.3 Call sequence diagram
+
```
Comment (etienne-lms) [inline documentation/benchmark.md]: I think the flow charts are fine but a bit confusing. Maybe a sentence should state the proposed setup: "**Benchmak CA** is used to setup the benchmark environment (allocation/registering of the benchmark shared buffers) and then to fork itself into target tested application with eventual CLI arguments. Once the target test application ends, timestamp data are made available in the generated `.ts` file(s)."

Comment (igoropaniuk) [inline documentation/benchmark.md]: I agree, but this information is already available in the **2.1 design overview** section

File: documentation/benchmark.md (hunk 20 above)
Comment (igoropaniuk) [inline documentation/benchmark.md]: My mistake, will fix it

Comment (igoropaniuk) [inline documentation/benchmark.md]: fixed

File: documentation/benchmark.md (hunk 21 above)
Comment (igoropaniuk) [inline documentation/benchmark.md]: @etienne-lms so what do you think, should this information be duplicated in the call sequence diagram?

Comment (etienne-lms) [inline documentation/benchmark.md]: I found it hard to understand how "CA benchmark" operate towards benchmark buffer registration and towards the target CA to be tested. Hence i though a short talk (description) could help figuring out the setup.

File: documentation/benchmark.md [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+provide detailed and precise profiling information for each layer.
+
+It is necessary to receive latency values for:
+* The roundtrip time for going from a client application in normal world,
+down to a Trusted Application and back again.
+* Detailed information for amount of time taken to go through each layer:
+	* libTEEC -> OP-TEE KMOD
+	* OP-TEE KMOD -> OP-TEE OS Core
+	* OP-TEE OS Core -> TA entry point (**not supported yet**)
+	* The same way back
```
Comment (etienne-lms) [inline documentation/benchmark.md]: Replace "OP-TEE KMOD" with "**Linux OP-TEE kernel driver**" for consistency.

File: documentation/benchmark.md [hunk 23]
```diff
@@ -0,0 +1,131 @@
...
+file with `.ts` extension. Additional details can be found here
+here [optee_benchmark]
+2. **Benchmark PTA**: pseudo TA, which owns all per-cpu circular non-secure
+buffers from a shared memory. **Benchmark PTA** must be invoked (by a CA)
+to register the timestamp circular buffers. In turn, the **Benchmark PTA**
+invokes the optee linux driver (through some RPC mean) to register this
+circular buffers in the linux kernel layer.
+3. **libTEEC** and **Linux OP-TEE kernel driver** include means to store
+timestamps in the circular buffers once these are registered by the
+**Benchmark pTA** into the **Linux OP-TEE kernel driver**.
```
Comment (etienne-lms) [inline documentation/benchmark.md]: Maybe replace the section "3." with something that state that `All OP-TEE components (libTEEC, Linux kernel driver, OP-TEE core and TA libutee (TEE Core API lib)) do include means to store timestamps in the circular buffers once these are registered`.

Comment (igoropaniuk) [inline documentation/benchmark.md]: I've re-grouped a bit, I left information in the **3** point that "these layers handler buffer registration requests from Benchmark PTA" (check `md`), and in the new paragraph there is information which applies to all layers.

File: documentation/benchmark.md (hunk 22 above)
Comment (igoropaniuk) [inline documentation/benchmark.md]: fixed

File: documentation/benchmark.md (hunk 21 above)
Comment (igoropaniuk) [inline documentation/benchmark.md]: done, could you please check if I understood you correctly

**Comment by etienne-lms** [comment]
Looks ok to me . Thanks. `Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

**Comment by igoropaniuk** [comment]
@etienne-lms thanks a lot for your review!

Rebased and applied tags

**Comment by igoropaniuk** [comment]
@jforissier this PR is ready for merge. Thanks!

**Pull request merged**